### PR TITLE
Implementation RenderStatePackager

### PR DESCRIPTION
**Pull request by MikhailGorobets**

### CODE REVIEW

File: RenderStatePackager/include/ParsingEnvironment.hpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+#include "ArchiverFactory.h"
+#include "ArchiverFactoryLoader.h"
+#include "RenderStatePackager.hpp"
+
+namespace Diligent
+{
+
+struct ParsingEnvironmentCreateInfo
+{
+    RENDER_DEVICE_TYPE_FLAGS DeviceBits;
```
Comment (TheMostDiligent) [inline RenderStatePackager/include/ParsingEnvironment.hpp:40]: Please always use default initializers for primitive types

File: RenderStatePackager/src/Main.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
+/*
+ *  Copyright 2019-2021 Diligent Graphics LLC
```
Comment (TheMostDiligent) [inline RenderStatePackager/src/Main.cpp:2]: It is better to name the file "main" (without capital M) to be consistent with standard c/c++ notation

File: RenderStatePackager/src/ParsingEnvironment.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    return m_pShaderStreamFactory;
+}
+
+RenderStatePackager* ParsingEnvironment::GetDeviceObjectConverter()
+{
+    return m_pDeviceReflection.get();
+}
+
+ParsingEnvironment::ParsingEnvironment(const ParsingEnvironmentCreateInfo& CreateInfo) :
+    m_CreateInfo(CreateInfo)
```
Comment (TheMostDiligent) [inline RenderStatePackager/src/ParsingEnvironment.cpp:54]: Prefer using brace initialization {}

File: RenderStatePackager/src/ParsingEnvironment.cpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+#else
+    m_pArchiveBuilderFactory = Diligent::GetArchiverFactory();
+#endif
+
+    SerializationDeviceCreateInfo                       DeviceCI = {};
+    RefCntAutoPtr<ISerializationDeviceDescriptorParser> pDesciptorParser;
+
+    if (!m_CreateInfo.ConfigFilePath.empty())
+    {
+        Diligent::CreateSerializationDeviceDescriptorParserFromFile(m_CreateInfo.ConfigFilePath.c_str(), &pDesciptorParser);
```
Comment (TheMostDiligent) [inline RenderStatePackager/src/ParsingEnvironment.cpp:70]: I don't think there is a need to have separate `SerializationDeviceDescriptorParser` object. Parsing device description is only needed for this tool. I don't see other components that may need it.

File: RenderStatePackagerLib/include/Comporators.hpp [hunk 5]
```diff
@@ -0,0 +1,211 @@
+/*
```
Comment (TheMostDiligent) [inline RenderStatePackagerLib/include/Comporators.hpp:1]: This needs to be moved to respective core headers

File: RenderStatePackagerLib/interface/DeviceObjectDescriptorParser.h [hunk 6]
```diff
@@ -0,0 +1,305 @@
+/*
```
Comment (TheMostDiligent) [inline RenderStatePackagerLib/interface/DeviceObjectDescriptorParser.h:1]: After another thought I suggest renaming this interface to IRenderStateMarkupParser and to name our json file format diligent render state markup (DRSM).

File: RenderStatePackagerLib/interface/DeviceObjectDescriptorParser.h [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+#pragma once
+
+ /// \file
+ /// Defines Diligent::IDeviceObjectDescriptorParser interface
+#include "../../../DiligentCore/Graphics/GraphicsEngine/interface/RenderDevice.h"
+
+DILIGENT_BEGIN_NAMESPACE(Diligent)
+
+
+struct GraphicsPipelineDescJSON
```
Comment (TheMostDiligent) [inline RenderStatePackagerLib/interface/DeviceObjectDescriptorParser.h:36]: `GraphicsPipelineMarkup`

File: RenderStatePackagerLib/interface/DeviceObjectDescriptorParser.h [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+    {   
+        //TODO
+        return false;
+    }
+#endif
+
+};
+typedef struct GraphicsPipelineDescJSON GraphicsPipelineDescJSON;
+
+struct PipelineStateCreateInfoJSON
```
Comment (TheMostDiligent) [inline RenderStatePackagerLib/interface/DeviceObjectDescriptorParser.h:51]: PipelineStateCreateMarkup

File: RenderStatePackagerLib/interface/DeviceObjectDescriptorParser.h [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+
+#if DILIGENT_CPP_INTERFACE
+    constexpr bool operator == (const PipelineStateCreateInfoJSON& RHS) const 
+    {
+        //TODO
+        return false;
+    }
+#endif  
+};
+typedef struct GraphicsPipelineDescJSON GraphicsPipelineDescJSON;
```
Comment (TheMostDiligent) [inline RenderStatePackagerLib/interface/DeviceObjectDescriptorParser.h:65]: GraphicsPipelineMarkup

File: RenderStatePackagerLib/interface/DeviceObjectDescriptorParser.h [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+
+struct GraphicsPipelineStateCreateInfoJSON DILIGENT_DERIVE(PipelineStateCreateInfoJSON)
```
Comment (TheMostDiligent) [inline RenderStatePackagerLib/interface/DeviceObjectDescriptorParser.h:67]: GraphicsPipelineStateCreateMarkup, etc.
Replace Info or Desc with Markup

File: RenderStatePackagerLib/interface/DeviceObjectDescriptorParser.h [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+#define DeviceObjectDescriptionParserInclusiveMethods \
+    IObjectInclusiveMethods;                          \
+    IDeviceObjectDescriptorParser DeviceObjectDescriptorParser
+
+
+DILIGENT_BEGIN_INTERFACE(IDeviceObjectDescriptorParser, IObject)
+{
+    VIRTUAL Bool METHOD(GetGraphicsPipelineState)(THIS_
+                                                  const Char* Name,
+                                                  GraphicsPipelineStateCreateInfo REF CreateInfo,
```
Comment (TheMostDiligent) [inline RenderStatePackagerLib/interface/DeviceObjectDescriptorParser.h:207]: Here and in other places I'd move GraphicsPipelineStateCreateInfo to Markup struct

File: RenderStatePackagerLib/interface/DeviceObjectDescriptorParser.h [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+    VIRTUAL Bool METHOD(GetShader)(THIS_
+                                   const Char* Name,
+                                   ShaderCreateInfo REF CreateInfo) CONST PURE;
+
+    VIRTUAL Bool METHOD(GetRenderPass)(THIS_
+                                       const Char* Name,
+                                       RenderPassDesc REF CreateInfo) CONST PURE;
+
+    
+    VIRTUAL Bool METHOD(EnumGraphicsPipelineStates)(THIS_
```
Comment (TheMostDiligent) [inline RenderStatePackagerLib/interface/DeviceObjectDescriptorParser.h:238]: This is not a very clear API. I suggest adding GetPipelineStateCount method() and
GetPipelineStateByIndex().
GetPipelineState can be renamed to GetPipelineStateByName to be consistent with shader reflection API

File: RenderStatePackagerLib/generated/DeviceObjectDescriptorParserParser.hpp [hunk 13]
```diff
@@ -0,0 +1,230 @@
+/*
```
Comment (TheMostDiligent) [inline RenderStatePackagerLib/generated/DeviceObjectDescriptorParserParser.hpp:1]: Why is there DeviceObjectDescriptorParser.hpp in both generated and include fodlers?

File: RenderStatePackagerLib/include/DeviceObjectDescriptorParserImpl.hpp [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+#include "RefCntAutoPtr.hpp"
+#include "ObjectBase.hpp"
+#include "DynamicLinearAllocator.hpp"
+#include "HashUtils.hpp"
+
+namespace Diligent
+{
+
+/// Implementation of IDeviceObjectDescriptorParser
+class DeviceObjectDescriptorParserImpl final : public ObjectBase<IDeviceObjectDescriptorParser>
```
Comment (TheMostDiligent) [inline RenderStatePackagerLib/include/DeviceObjectDescriptorParserImpl.hpp:37]: RenderStateMarkupParserImpl

File: RenderStatePackagerLib/interface/SerializationDeviceDescriptorParser.h [hunk 15]
```diff
@@ -0,0 +1,73 @@
+/*
```
Comment (TheMostDiligent) [inline RenderStatePackagerLib/interface/SerializationDeviceDescriptorParser.h:1]: There seems to be no need for this interface. It is specific for RenderStatePackager and is not used anywhere else.

**Review by TheMostDiligent — CHANGES_REQUESTED** [review]

File: RenderStatePackagerLib/generated/DeviceObjectDescriptorParserParser.hpp (hunk 13 above)
Comment (MikhailGorobets) [reply RenderStatePackagerLib/generated/DeviceObjectDescriptorParserParser.hpp:1]: This is `DeviceObjectDescriptorParserParser.hpp`.  This file has structures for which we generate `Serialize` and `Deserialize`

**Comment by TheMostDiligent** [comment]
Merged into master in https://github.com/DiligentGraphics/DiligentTools/commit/837a5d5ca67f92b59beec98a0c0e4aa9cad3bbc7

**Pull request closed without merging**